### ISSUE TITLE
MLE gives incorrect value with Uniform RVs

### ISSUE THREAD

**Issue by kyleabeauchamp** [pymc-devs/pymc#2482]
I'm finding that MAP / MLE estimates with Uniform RVs are not converging with the level of precision that I need.  I've found a very simple test case where a 1D normal ends up being off by > 10% of the standard deviation.  

Coding up the same (trivial) example in numpy+scipy gets excellent convergence.

Switching the Uniform prior to a flat prior seems to solve the problem, so I conclude that it's either:

1.  LogP terms leaking into the model from the uniform variable transformations
2.  Major precision problems in the logP and gradient associated with the uniform variable transformations

I have already tried multiple scipy solvers and convergence settings, so IMHO that is not the issue.

Example below:

```
import pymc3 as pm
import numpy as np
import scipy.optimize
from scipy import special


x0 = 10.0
sd = 1.0
a, b = (9, 12)  # bounds for uniform RV, need non-symmetric to reproduce issue

with pm.Model() as model1:
    mu_i = pm.Flat("mu_i")
    x = pm.Normal("X_obs", mu=mu_i, sd=sd, observed=x0)

map_estimate1 = pm.find_MAP(model=model1)
mu1 = map_estimate1["mu_i"]


with pm.Model() as model2:
    mu_i = pm.Uniform("mu_i", a, b)
    x = pm.Normal("X_obs", mu=mu_i, sd=sd, observed=x0)

map_estimate2 = pm.find_MAP(model=model2)
# Manually invert the transformation; I've also done this automatically with PR #2468 with same results
mu2 = (b - a) * special.expit(map_estimate2["mu_i_interval__"]) + a


def objective(mu):
    return -0.5 * np.log(2 * np.pi) + 0.5 * (x0 - mu) ** 2


def gradient(mu):
    return -1 * (x0 - mu)


ans = scipy.optimize.minimize(objective, 0.0, jac=gradient, method='L-BFGS-B')
mu3 = ans["x"]

print("mu1, mu2, mu3")
print(mu1, mu2, mu3)
10.0 10.2391245943 [ 10.]
```

*edit fixed manual inverse transform.

**Comment by aseyboldt** [comment]
First, I think you inverse transform is wrong, it should be
```python
from scipy import special
(b - a) * special.expit(map_estimate2["mu_i_interval__"]) + a
```
But this doesn't solve the problem, the MAP is still not 10.

The problem here is that the MAP is not invariant under reparametrization (the MAP is not a property of the posterior distribution, but a property of the density function), but our implementation of the transformed variables assumes that inference algorithms are. So I think I'd consider this a bug in pymc3.

A solution to the problem would be to add an option when accessing the logp that disables the jacobians of the transformations, and use that raw logp as objective in the MAP.

**Comment by kyleabeauchamp** [comment]
Yeah, I didn't know the internals but my hunch was that a jacobian term was being applied as part of the MAP potential.  

I think MAP / MLE estimation is worth supporting, as they're very often the first approach used in the literature, so being able to quickly code them up in pymc3 is a big advantage--even if the eventual goal is to migrate towards something more sophisticated.

**Comment by kyleabeauchamp** [comment]
I glanced at the various distribution classes but I didn't see an obvious way to check whether each variable is a jacobian term or not...

**Comment by junpenglao** [comment]
Yeah it is definitely b/c of the transform. Actually if you set `transform=None` in the `mu_i`, running `find_MAP` gives the correct output.

**Comment by junpenglao** [comment]
@kyleabeauchamp the Jacobian is added in the transform automatically.

**Comment by aseyboldt** [comment]
@kyleabeauchamp The jacobians are hidden `distributions/transforms.py`

**Comment by kyleabeauchamp** [comment]
Is there a description somewhere of what it means to set transform=None?  Is that disabling the bounds?  Implementing them another way?

**Comment by aseyboldt** [comment]
Sadly, transforms aren't exactly well documented at moment...
Setting transform=None will prevent pymc3 from creating one of those transformed `varname_<transformname>__` variables, and it will work on the raw values instead. The bounds are still enforced by returning -inf for values outside the bounds, but optimizers won't be particularly happy about that. I think transformed vars without the jacobian term are the right solution.

**Comment by junpenglao** [comment]
I think the current implementation should be correct:

Without adjusting for Jacobian:
```python
with pm.Model() as model1:
    mu_i = pm.Flat("mu_i")
    theta = pm.Deterministic('theta', pm.math.invlogit(mu_i))
    x = pm.Normal("X_obs", mu=theta*3 + 9., sd=sd, observed=x0)
    #constrain = pm.Potential('constrain', tt.log(theta)+tt.log(1-theta))

map_estimate1 = pm.find_MAP(model=model1)
print(pm.math.invlogit(map_estimate1['mu_i']).eval()*3 + 9)
```
```python
10.0000001079
```

Correctly adjusting for Jacobian:
```python
with pm.Model() as model1:
    mu_i = pm.Flat("mu_i")
    theta = pm.Deterministic('theta', pm.math.invlogit(mu_i))
    x = pm.Normal("X_obs", mu=theta*3 + 9., sd=sd, observed=x0)
    constrain = pm.Potential('constrain', tt.log(theta)+tt.log(1-theta))

map_estimate1 = pm.find_MAP(model=model1)
print(pm.math.invlogit(map_estimate1['mu_i']).eval()*3 + 9)
```
```python
10.2391245943
```

**Comment by junpenglao** [comment]
Also, FYI: http://mc-stan.org/users/documentation/case-studies/mle-params.html

**Comment by kyleabeauchamp** [comment]
I understand the non-invariant behavior of MLE quantities, but I still think they are worth having.

Also, I think @aseyboldt and I are claiming that the MLE should not include the extra Jacobian term as a potential.

**Comment by kyleabeauchamp** [comment]
In particular, the models I'm comparing here have priors that are constant, so IMHO there should not be any accumulation of Jacobian terms.

**Comment by aseyboldt** [comment]
I agree. Dislike for an inference method isn't reason to return wrong results ;-)
I'll write a patch tomorrow.

On August 7, 2017 7:43:35 PM CEST, Kyle Beauchamp <notifications@github.com> wrote:
>In particular, the models I'm comparing here have priors that are
>constant, so IMHO there should not be any accumulation of Jacobian
>terms.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pymc-devs/pymc3/issues/2482#issuecomment-320731345

-Adrian

**Comment by kyleabeauchamp** [comment]
One last note: working off of #2468 made it quite a bit easier to diagnose this issue because I was able to trivially compare the estimates made using priors with different underlying transforms.

**Comment by kyleabeauchamp** [comment]
Anyone else have any thoughts on the easiest path to fix this bug?

**Comment by aseyboldt** [comment]
I just started fixing this when I saw your earlier post about tensorflow :-)
Sorry for taking so long.

**Comment by kyleabeauchamp** [comment]
No worries, just wanted to double check before I started making a foolhardy attempt at fixing it...

**Comment by kyleabeauchamp** [comment]
Fixed in #2499
